kernel/src/engine/arrow_utils.rs

```diff
@@ -66,17 +69,63 @@ pub(crate) fn make_arrow_error(s: impl Into<String>) -> Error {
     .with_backtrace()
 }
 
+/// Prepares to enumerate row indexes of rows in a parquet file, accounting for row group skipping.
+pub(crate) struct RowIndexBuilder {
+    row_group_starting_row_offsets: Vec<Range<i64>>,
+    row_group_ordinals: Option<Vec<usize>>,
+}
+impl RowIndexBuilder {
+    pub(crate) fn new(row_groups: &[RowGroupMetaData]) -> Self {
+        let mut row_group_starting_row_offsets = vec![];
+        let mut offset = 0;
+        for row_group in row_groups {
+            let num_rows = row_group.num_rows();
+            row_group_starting_row_offsets.push(offset..offset + num_rows);
+            offset += num_rows;
+        }
+        Self {
+            row_group_starting_row_offsets,
+            row_group_ordinals: None,
+        }
+    }
+
+    /// Only produce row indexes for the row groups specified by the ordinals that survived row
+    /// group skipping. The ordinals must be in 0..num_row_groups.
+    pub(crate) fn select_row_groups(&mut self, ordinals: &[usize]) {
+        // NOTE: Don't apply the filtering until we actually build the iterator, because the
+        // filtering is not idempotent and `with_row_groups` could be called more than once.
+        self.row_group_ordinals = Some(ordinals.to_vec())
+    }
+}
+
+impl IntoIterator for RowIndexBuilder {
+    type Item = i64;
+    type IntoIter = std::iter::Flatten<std::vec::IntoIter<Range<Self::Item>>>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        let starting_offsets = match self.row_group_ordinals {
+            Some(ordinals) => ordinals
+                .iter()
+                .map(|i| self.row_group_starting_row_offsets[*i].clone())
+                .collect(),
+            None => self.row_group_starting_row_offsets,
+        };
+        starting_offsets.into_iter().flatten()
```

Comment:
We could avoid the clone with:
```suggestion
        if let Some(ordinals) = self.row_group_ordinals {
            let mut keep = vec![false; self.row_group_starting_row_offsets.len()];
            for i in ordinals.iter() {
                keep[*i] = true;
            }
            let mut iter = keep.iter();
            self.row_group_starting_row_offsets.retain(|_| *iter.next().unwrap());
        }
        self.row_group_starting_row_offsets.into_iter().flatten()
```

Trades against allocating the vec, so not sure it's actually much better.

Reply:
Clone of a `Range<usize>` should be super cheap -- no worse than a fat pointer. IMO it should be Copy, but 🤷